### PR TITLE
Automatic version bumping

### LINKED ISSUE #94: Automate bumping of version in app/build.gradle
Goal: Make it possible to automatically bump the `versionCode` and `versionName`  in [`app/build.gradle`](https://github.com/elimu-ai/vitabu/blob/main/app/build.gradle) when releasing a new APK version.

It should work similarly to the fully automated Maven release versioning setup that we have already implemented in the [webapp](https://github.com/elimu-ai/webapp/blob/main/pom.xml#L134): 
- `mvn release:clean release:prepare release:perform`
  - Sample: "[[maven-release-plugin] prepare release webapp-2.2.118](https://github.com/elimu-ai/webapp/commit/05f65d229d2b1389a8cdaaefbf93c8b3469bc2ce)"
  - Sample: "[[maven-release-plugin] prepare for next development iteration](https://github.com/elimu-ai/webapp/commit/7e19a82e3ac46280fdb8fa9753ad35a0b66e61c6)"

This way, an APK release of the Android app can be triggered with a Gradle command with a CI tool (e.g. Jenkins). So just like the `<version>` in the webapp's `pom.xml` gets automatically bumped, we would like to do the same with the Android app's `versionCode` and `versionName`.

### PR DESCRIPTION
**Pull request by hhio618**
### Description
closes #94 
### Tasks
1. Ensure clean repo and bumping a release (using either type or version parameters) and commit the changes and prepare for the next development cycle:
```
$ ./gradlew releasePrepare [-PbumpType=(major|minor|patch)] [-PbumpVersion=x.y.z]
```
2. Push the changes to git using the Gradle-git plugin:
```
$ ./gradlew releasePerform
```

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -7,7 +7,7 @@ buildscript {
     }
     dependencies {
         classpath 'com.android.tools.build:gradle:7.1.2'
-        
+        classpath 'org.ajoberstar:gradle-git:+'
```
Comment (nya-elimuai) [inline build.gradle:10]: @hhio618 Would it be better if we put a specific version of the library?

**Review by nya-elimuai — CHANGES_REQUESTED** [review]
@hhio618 Thank you, this is really cool to see!

I tried running `./gradlew releasePrepare`, and got "unspecified" as the new versionName:

```
Bumping versionName from 1.1.4 to unspecified
Bumping versionCode from 1001004 to 1001005
```

Screen Shot 2022-05-21 at 7 02 41 PM

Would it be possible to set the `versionName` as well?

**Comment by hhio618** [comment]
Thanks, @nya-elimuai. you need to specify either `bumpType` or `bumpVersion` parameters to get this to work:
Example:
```
./gradlew releasePrepare -PbumpType=patch # will increase the patch number, choices=major,minor,patch
```
```
./gradlew releasePrepare -PbumpVersion=x.y.z # bump to a specific version.
```
I added an exception to prompt for either of these parameters.

File: app/build.gradle [hunk 2]
```diff
@@ -85,11 +85,12 @@ task releasePrepare(dependsOn: releaseClean) {
 
         // Set new version name from input parameters.
         def newVersionName
-        if (project.properties.containsKey("version")){
-            newVersionName = project.properties["version"]
-        }else {
+        print project.properties.containsKey("bumpVersion")
+        if (project.properties.containsKey("bumpVersion")){
```
Comment (nya-elimuai) [inline app/build.gradle:89]: Add missing whitespace: `) {`

**Comment by nya-elimuai** [comment]
> https://github.com/elimu-ai/vitabu/pull/96#issuecomment-1133643779

@hhio618 This is really cool! Tested with `./gradlew releasePrepare -PbumpType=patch`:

Screen Shot 2022-05-27 at 8 08 06 PM

Screen Shot 2022-05-27 at 8 08 14 PM

**Review by nya-elimuai — CHANGES_REQUESTED** [review]
@hhio618 This is really great! Amazing work 👏🏽

I only have one final adjustment, when running `./gradlew releasePrepare -PbumpType=patch`: Instead of bumping the version _before_ the release, bump it _after_ the release. That way it would work the same as with the webapp release versioning.

Example: If the version looks like this before the release:
```gradle
versionCode 1001005
versionName '1.1.5-SNAPSHOT'
```
change it to 
```gradle
versionCode 1001005
versionName '1.1.5'
```

Then release (commit to Git).


And then (after the release), bump the version for the next iteration:
```gradle
versionCode 1001006
versionName '1.1.6-SNAPSHOT'
```

**Comment by hhio618** [comment]
@nya-elimuai Thanks for the review! I made the changes you mentioned.

File: app/build.gradle [hunk 3]
```diff
@@ -61,22 +61,48 @@ dependencies {
...
     doLast {
         if (!grgit.repository.jgit.status().call().clean) {
-            throw new GradleException('Git status is not clean, please commit or stash your changes!')
+            throw new GradleException('Git status is not clean, please stash your changes!')
         }
     }
 }
+task releaseClean(dependsOn: ensureCleanRepo) {
+    doLast {
+        def Clean = true
```
Comment (nya-elimuai) [inline app/build.gradle:73]: Make `Clean` lower-case: `clean`

File: app/build.gradle [hunk 4]
```diff
@@ -61,22 +61,48 @@ dependencies {
...
+        String headCommitMessage = grgit.head().shortMessage
+        while (headCommitMessage.contains("[gradle-release-task]")) {
+            Clean = false
+            println "Found git commit: $headCommitMessage"
+            if (headCommitMessage.indexOf("vitabu-") > -1) {
```
Comment (nya-elimuai) [inline app/build.gradle:78]: To make the solution scalable to other Android apps (with different names, can you change "vitabu-" into a variable? For example by using the `applicationId` variable instead?

File: app/build.gradle [hunk 5]
```diff
@@ -61,22 +61,48 @@ dependencies {
...
+                def tagName = headCommitMessage.split("vitabu-")[1]
```
Comment (nya-elimuai) [inline app/build.gradle:79]: Same comment here as above.

File: app/build.gradle [hunk 6]
```diff
@@ -61,22 +61,48 @@ dependencies {
...
+                } catch (Exception e) {
+                    println "Error while removing git tag:\n $e"
+                }
+            }
+            println "Resetting the git commit permanently!"
+            grgit.reset(commit: "HEAD~1", mode: "hard")
+            headCommitMessage = grgit.head().shortMessage
+
+        }
+        if (Clean){
```
Comment (nya-elimuai) [inline app/build.gradle:94]: Change to `if (clean) {`

**Review by nya-elimuai — CHANGES_REQUESTED** [review]
Looks great! Only added a suggestion to use the `applicationId` variable.

**Review by nya-elimuai — APPROVED** [review]
Looks good. Thank you!

**Pull request merged**